src/EditDialog.cpp

```diff
@@ -84,59 +82,166 @@ void EditDialog::reject()
+            // Clear any image from the image viewing widget
+            ui->editorImage->setPixmap(QPixmap(0,0));
+
+            // Load the Null into the hex editor
+            hexEdit->setData(data);
+
+            // The hex widget buffer is now the main data source
+            dataSource = HexBuffer;
+        }
+        break;
 
-    // Do some data type specific handling
-    QImage img;
-    switch (dataType) {
     case Text:
-        // If in text editor mode, select all of the text by default
-        if (editMode == 0) {
+        switch (editMode) {
+        case TextEditor:
+            // Load the text into the text editor
+            textData = QString::fromUtf8(data.constData(), data.size());
+            ui->editorText->setPlainText(textData);
+
+            // Enable text editing
+            ui->editorText->setEnabled(true);
+
+            // Select all of the text by default
             ui->editorText->selectAll();
-        }
 
-        // Clear any image from the image viewing widget
-        ui->editorImage->setPixmap(QPixmap(0,0));
+            // The text widget buffer is now the main data source
+            dataSource = TextBuffer;
+            break;
+
+        case HexEditor:
+            // Load the text into the hex editor
+            hexEdit->setData(data);
+
+            // The hex widget buffer is now the main data source
+            dataSource = HexBuffer;
+            break;
+
+        case ImageViewer:
+            // Clear any image from the image viewing widget
+            ui->editorImage->setPixmap(QPixmap(0,0));
+
+            // Load the text into the text editor
+            textData = QString::fromUtf8(data.constData(), data.size());
+            ui->editorText->setPlainText(textData);
+
+            // Enable text editing
+            ui->editorText->setEnabled(true);
+
+            // Select all of the text by default
+            ui->editorText->selectAll();
+
+            // The text widget buffer is now the main data source
+            dataSource = TextBuffer;
+            break;
+        }
         break;
 
     case Image:
-        // For images, load the image into the image viewing widget
-        if (img.loadFromData(data)) {
-            ui->editorImage->setPixmap(QPixmap::fromImage(img));
+        // Image data is kept in the hex widget, mainly for safety.  If we
+        // stored it in the editorImage widget instead, it would be a pixmap
+        // and there's no good way to restore that back to the original
+        // (pristine) image data.  eg image metadata would be lost
+        hexEdit->setData(data);
+
+        // The hex widget buffer is now the main data source
+        dataSource = HexBuffer;
+
+        // Update the display if in text edit or image viewer mode
+        switch (editMode) {
+        case TextEditor:
+            // Disable text editing, and use a warning message as the contents
+            ui->editorText->setText(QString("<i>" %
```

Comment:
Yep.  The link above explains that `+` is less efficient than `%` for QString concatenation, so I was trying to go with the better approach.

If that's a hassle, I'm ok to use `+` instead. :smile: